### PR TITLE
Update Jax to `0.3.15` for fixing general_param_shift demo

### PR DESCRIPTION
**Pull request by rmoyard**
**Changes:**

Update `jax`, `jax-lib` and `neural-tangents` to the latest version.

### CODE REVIEW

**Comment by rmoyard** [comment]
The build dev check: https://github.com/PennyLaneAI/qml/actions/runs/2782858608

File: .github/workflows/build-pr.yml [hunk 1]
```diff
@@ -7,7 +7,7 @@ on:
       - dev
 
 env:
-  NUM_WORKERS: 15
+  NUM_WORKERS: 14
```
Comment (antalszava) [inline .github/workflows/build-pr.yml:10]: How come this was changed?

Comment (rmoyard) [reply .github/workflows/build-pr.yml:10]: I talked with @rashidnhm and it is the hacky way to trigger all bbuild without cache.

Comment (rashidnhm) [reply .github/workflows/build-pr.yml:10]: Yes, currently there is not built-in way to ignore the cache for all demos. Changing this value invalidates the cache for your branch.

I will be adding a feature where changes to the requirements-*.txt files builds all demos fresh.

File: demonstrations/tutorial_ml_classical_shadows.py [hunk 2]
```diff
@@ -590,7 +590,7 @@ def build_dataset(num_points, Nr, Nc, T=500):
 
 for i in range(len(kernel_NN)):
     for j in range(len(kernel_NN)):
-        kernel_NN[i][j] /= (kernel_NN[i][i] * kernel_NN[j][j]) ** 0.5
+        kernel_NN.at[i, j].set((kernel_NN[i][i] * kernel_NN[j][j]) ** 0.5)
```
Comment (antalszava) [inline demonstrations/tutorial_ml_classical_shadows.py:593]: Why was this changed?

**Review by antalszava — APPROVED** [review]
Looks good :+1:

**Comment by rmoyard** [comment]
[sc-24264]
